sdk/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Upgrade.hs

```diff
@@ -53,19 +55,32 @@ instance Applicative Upgrading where
 foldU :: (a -> a -> b) -> Upgrading a -> b
 foldU f u = f (_past u) (_present u)
 
+unsafeFlattenUpgrading :: Upgrading [a] -> [Upgrading a]
+unsafeFlattenUpgrading = foldU (zipWith Upgrading)
+
 -- Allows us to split the world into upgraded and non-upgraded
-type TcUpgradeM = TcMF (Upgrading Gamma)
+type TcUpgradeM = TcMF UpgradingEnv
+
+data UpgradingEnv = UpgradingEnv
```

Comment:
I opted for my approach because `Upgrading Env` allows the `_deps` fields to have different lengths, whereas my type signature has `_upgradingDeps :: [Upgrading LF.PackageId]`, which guarantees that packageId's are paired. This makes the functions like `checkImport` a bit easier and nicer in my opinion.